File: packages/java/parser-jvm-core/src/main/java/com/vaadin/fusion/parser/core/PluginManager.java
package com.vaadin.fusion.parser.core;

import java.util.Collection;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

final class PluginManager {
    private static final Logger logger = LoggerFactory
            .getLogger(PluginManager.class);

    private final Set<Plugin> plugins;

    PluginManager(ParserConfig config) {
        plugins = config.getPlugins();
    }

    public void execute(Collection<RelativeClassInfo> endpoints,
            Collection<RelativeClassInfo> entities, SharedStorage storage) {
        for (var plugin : plugins) {
            logger.debug("Executing plugin " + plugin.getClass().getName());
            plugin.execute(endpoints, entities, storage);
        }
    }
}

File: packages/ts/generator-core/src/PluginManager.ts
import type LoggerFactory from '@vaadin/hilla-generator-utils/LoggerFactory.js';
import type Plugin from './Plugin.js';
import type { PluginConstructor } from './Plugin.js';
import type ReferenceResolver from './ReferenceResolver.js';
import type SharedStorage from './SharedStorage.js';

export default class PluginManager {
  readonly #plugins: Plugin[];

  constructor(plugins: readonly PluginConstructor[], resolver: ReferenceResolver, logger: LoggerFactory) {
    const standardPlugins = [
      'BackbonePlugin',
      'ClientPlugin',
      'BarrelPlugin',
      'ModelPlugin',
      'PushPlugin',
      'SubTypesPlugin',
    ];
    const customPlugins = plugins.filter((p) => !standardPlugins.includes(p.name));
    if (customPlugins.length > 0) {
      logger.global.info(`Plugins: ${plugins.map(({ name }) => name).join(', ')}`);
    }
    this.#plugins = plugins.map((PluginClass) => new PluginClass(resolver, logger));
  }

  async execute(storage: SharedStorage): Promise<void> {
    // We need to run plugins sequentially
    for (const plugin of this.#plugins) {
      const { name, path } = plugin;
      plugin.logger.debug({ plugin: { name, path } }, `Executing plugin '${plugin.name}'`);
      // eslint-disable-next-line no-await-in-loop
      await plugin.execute(storage);
    }
  }
}
